text/0049-parse-multiply-by-constant.md

```diff
@@ -0,0 +1,63 @@
+# Refine the Parsing Rule of Multiply-By-Constant Operation
+
+## Related issues and PRs
+
+- Reference Issues: https://github.com/cedar-policy/cedar/issues/619
+- Implementation PR(s): (leave this empty)
+
+## Timeline
+
+- Started: 2024-02-07
+
+## Summary
+
+This PR proposes that the parser checks validity of multiplication operations after left-associativity is explicitly applied on the CST level. And the order of resulting AST should be the same as that of CST.
+
+## Basic example
+
+`2*1*context.value` should be conceptually converted into `((2*1)*context.value)` before checking if it's a valid multiply-by-constant operation. The checking should pass and the resulting AST should be `(mul (mul 2 1) context.value)`.
+
+## Motivation
+
+Let's first revisit the status quo: The parser separates the operands of a CST representing a (possibly chained) multiplication operation into a non-integer-constant and a list of integer constants, preserving the order of the constants, and then *left-fold* the constant list with the non-constant. For instance, `2*1*content.value` is parsed into `(mul (mul content.value 2) 1)`.
+
+Note that the parser changes the order of operands, which does not alter the evaluation result since Cedar's multiplication operation is associative. However, changing the evaluation order can make error message confusing: For instance, evaluating `2*6*context.value` when `context.value` is `i64::MAX/10` produces an error message like ```integer overflow while attempting to multiply `1844674407370955160` by `6```, contradicting the intuitive left-to-right evaluation order.
```

Comment:
This isn't true. Consider the policy:
```
permit(principal, action, resource) when
{
    6*0*principal.foo >= 0
};
```
Where `User::"Bob".foo` = 1844674407370955160.
Authorization will deny due to overflow. If we change the policy to:
```
permit(principal, action, resource) when
{
    0*6*principal.foo >= 0
};
```
Authorization will allow.